File: Task2/src/main/java/net/omsu/imit/course3/VectorOperation.java
package net.omsu.imit.course3;

public class VectorOperation {

    public static double[] multiplication(SquareMatrix mtrx,double[] x){
        int N=mtrx.getSize();
        double[] res=new double[N];
        double t;

        for(int i=0;i<N;i++){
            t=0;
            for(int j=0;j<N;j++){
                t+=mtrx.getElem(i,j)*x[j];
            }
            res[i]=t;
        }
        return res;
    }

    public static double[] subtraction(double[]x,double[] y){
        int N=x.length;
        double[] res=new double[N];

        for(int i=0;i<N;i++){
            res[i]=x[i]-y[i];
        }
        return res;
    }

    public static double[] addition(double[]x,double[] y){
        int N=x.length;
        double[] res=new double[N];

        for(int i=0;i<N;i++){
            res[i]=x[i]+y[i];
        }
        return res;
    }

    public static double[] multiplyByNumber(double[] x,double t){
        double[] res=new double[x.length];

        for(int i=0;i<x.length;i++){
           res[i]=x[i]*t;
        }
        return res;
    }

    public static void print(double[]x){
        for (int i = 0; i < x.length; i++) {
            System.out.print(x[i] + " ");
        }
    }
}

File: Task2_cpp_version/VectorOperation.cpp
#include <iostream>
#include "VectorOperation.h"
#include <cmath>
using namespace std;
#define PI 3.141592653589793238462643383279502884

double* createVector(int n) {
	double* vector = new double[n];
	for (int i = 0; i < n; i++) {
		vector[i] = 0;
	}
	return vector;
}

void multiplication(double **mtrx, double *&X, int N) {
	double* t=createVector(N);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			t[i]+=mtrx[i][j]*X[j];
		}
	}
	X = t;
}

void multiplyByNumber(double *& X, double tau,int N)
{
	for (int i = 0; i < N; i++) {
		X[i] *= tau;
	}
}

void subtraction(double* F, double *& X, int N)
{
	for (int i = 0; i < N; i++) {
			X[i] = F[i]-X[i];
	}
}

void addition(double *& oldX, double*& newX, int N)
{
	for (int i = 0; i < N; i++) {
		newX[i] += oldX[i];
		oldX[i] = newX[i];
	}
}

void printVector(double*X,int N) {
	for (int i = 0; i < N; i++) {
		cout<<X[i]<<" ";
	}
	cout << endl;
}

double* copyVector(double*X,int N) {
	double*t =createVector(N);
	for (int i = 0; i < N; i++) {
		t[i] += X[i];
	}
	return t;
}


